pandas/core/indexes/range.py

```diff
@@ -433,8 +436,19 @@ def delete(self, loc) -> Int64Index:  # type: ignore[override]
 
     def take(
         self, indices, axis: int = 0, allow_fill: bool = True, fill_value=None, **kwargs
-    ) -> Int64Index:
+    ) -> Int64Index | RangeIndex:
         with rewrite_exception("Int64Index", type(self).__name__):
+            if (
+                fill_value is None
+                and isinstance(indices, np.ndarray)
```

Comment:
this is way too complicated and likely hiding bugs. @jbrockmendel can you suggest anything here.

Reply:
cc @phofl